compiler/typecheck/TcDaml.hs

```diff
@@ -13,34 +11,310 @@ import OccName
+  foldMap extractDamlInfoFromTyThing (typeEnvElts $ eps_PTE eps) `mappend`
+    foldMap extractDamlInfoFromClsInst (instEnvElts $ eps_inst_env eps)
+
+extractDamlInfoFromTyThing :: TyThing -> DamlInfo
+extractDamlInfoFromTyThing tything =
+  case tything of
+    AnId id ->
+      mempty
+        { methods = mapMaybe matchMethod [id]
+        , implementations = mapMaybe matchImplements [id]
+        , choices = mapMaybe matchChoice [id]
+        }
+    ATyCon tycon ->
+      mempty
+        { templates = mapMaybe matchTemplate [tycon]
+        , interfaces = mapMaybe matchInterface [tycon]
+        }
+    _ -> mempty
+  where
+    matchTemplate, matchInterface :: TyCon -> Maybe Name
+    matchTemplate = tyconWithConstraint "DamlTemplate"
+    matchInterface = tyconWithConstraint "DamlInterface"
+
+    tyconWithConstraint :: String -> TyCon -> Maybe Name
+    tyconWithConstraint targetName tycon
+      | isAlgTyCon tycon
+      , let isMatchingLoneConstraint type_
+              | Just (loneConstraint, []) <- splitTyConApp_maybe type_
+              , similarName targetName (tyConName loneConstraint)
+              = True
+              | otherwise
+              = False
+      , any isMatchingLoneConstraint (tyConStupidTheta tycon)
+      = Just $ tyConName tycon
+      | otherwise
+      = Nothing
+
+    matchChoice :: Id -> Maybe (Name, Name)
+    matchChoice identifier
+      | TyConApp headTyCon [contractType, choiceType, returnType] <- varType identifier
+      , similarName "HasExercise" (tyConName headTyCon)
+      , Just (contractTyCon, []) <- splitTyConApp_maybe contractType
+      , Just (choiceTyCon, []) <- splitTyConApp_maybe choiceType
+      = Just (tyConName contractTyCon, tyConName choiceTyCon)
+      | otherwise
+      = Nothing
+
+    matchMethod :: Id -> Maybe (FastString, (Name, Type))
+    matchMethod identifier
+      | TyConApp headTyCon [contractType, methodNameType, returnType] <- varType identifier
+      , similarName "HasMethod" (tyConName headTyCon)
+      , LitTy (StrTyLit methodName) <- methodNameType
+      , Just (contractTyCon, []) <- splitTyConApp_maybe contractType
+      = Just (methodName, (tyConName contractTyCon, returnType))
+      | otherwise
+      = Nothing
+
+    matchImplements :: Id -> Maybe (Name, Name)
+    matchImplements identifier
+      | TyConApp headTyCon [templateType, interfaceType] <- varType identifier
+      , similarName "HasToInterface" (tyConName headTyCon)
+      , Just (templateTyCon, []) <- splitTyConApp_maybe templateType
+      , Just (interfaceTyCon, []) <- splitTyConApp_maybe interfaceType
+      = Just (tyConName templateTyCon, tyConName interfaceTyCon)
+      | otherwise
+      = Nothing
+
+    isMatchingLoneConstraint :: String -> IfaceType -> Bool
+    isMatchingLoneConstraint targetName type_
+      | IfaceTyConApp (IfaceTyCon tyConName _info) IA_Nil <- type_
+      , similarName targetName tyConName
+      = True
+      | otherwise
+      = False
+
```

Comment:
This appears to be unused. There's a similar local function with the same name under `tyconWithConstraint`, but that one doesn't take an `IfaceType`

```suggestion
```

Reply:
removed